core/src/main/java/jenkins/model/Jenkins.java

```diff
@@ -4976,14 +4993,49 @@ private static void computeVersion(ServletContext context) {
      * Version number of this Jenkins.
      */
     public static String VERSION = UNCOMPUTED_VERSION;
-
+    
+    /**
+     * Minimal Java level required to run this Jenkins.
+     * If {@code null}, the version is unknown.
+     */
+    private static @CheckForNull String JAVA_MIN_LEVEL = null;
+    private static VersionNumber DEFAULT_JAVA_MIN_LEVEL = new VersionNumber("8");
```

Comment:
> I don't understand what's going on here. Is all of this to support a custom (higher) minimum release is repackaged Jenkins distributions?

AFAIK there is no custom distributions requiring it, but plugin maintainers may bump Java level beyond the core. We had it for Yet Another Docker plugin and several else. If somebody uses Java 9 plugins in the future, they can mark the requirement as a system property